Refactor cfstore to dataclasses

jacomago opened the pull request:
Large refactor of cfstore

- Adds dataclasses to handle all the intemediary types rather than the usage of dictionaries for everything.
- Swaps to snake case (where not using Twisted stuff)
- Adds typing everywhere I can
- Adds pydoc to every method
- Splits up the bigger methods to be much smaller
- Adds the full data structures to the log messages
- Cleans up exception messages to include more debug information

jacomago committed:
Use a dataclass for CFConfig

jacomago committed:
Use dataclass for CFProperty

jacomago committed:
Add committransaction dataclass

jacomago committed:
Add RecordInfo dataclass

jacomago committed:
Add CFPropertyName enum

jacomago committed:
Add PVStatus enum

jacomago committed:
Create IocInfo Dataclass

jacomago committed:
Add CFChannel dataclass

jacomago committed:
sleep is a float

jacomago committed:
CancelledError has no host or port properties

jacomago committed:
Split up commitWithThread

jacomago committed:
Improve logging

log the whole object when available
use inbuilt log formatting rather than f strings or .format (should improve performance)

jacomago committed:
Rename create_properties to create_ioc_properties

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -103,25 +135,25 @@ def _startServiceWithLock(self):
                 if infotags_whitelist:
                     record_property_names_list = [s.strip(", ") for s in infotags_whitelist.split()]
                 else:
                     record_property_names_list = []
```
Do you need this conditional block now? The default is an empty string, so 
```py
record_property_names_list = [s.strip(", ") for s in infotags_whitelist.split()]
```
should return an empty list in that case.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -39,16 +37,50 @@
+    username: str = "cfstore"
+    recceiverId: str = RECCEIVERID_DEFAULT
+    timezone: str = ""
+    findSizeLimit: int = 10000
```
cAmElCaSe vs s_n_a_k_e_c_a_s_e?

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -735,7 +738,10 @@ def __updateCF__(
+            {"name": ch["name"], "owner": ch["owner"], "properties": ch["properties"].as_dict()}
+            for ch in channels[i : i + chunk_size]
+        ]
         client.set(channels=chunk)
```
Some type hints here would be nice. It's not clear what `channels` is.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -110,7 +117,7 @@ def _startServiceWithLock(self):
             self.client = ChannelFinderClient()
             try:
-                cf_properties = [cf_property["name"] for cf_property in self.client.getAllProperties()]
+                cf_properties = {cf_property["name"] for cf_property in self.client.getAllProperties()}
```
I see that this is used as a `set` below; if you're making this change, maybe remove the `set()`s below?

simon-ess replied:
Also, you're creating a CFProperty data class here but here you're reading things as a dictionary. This commit is a bit confusing.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -74,7 +65,10 @@ def load_from_conf(conf) -> "CFConfig":
+    value: Optional[str] = None
+
+    def as_dict(self) -> Dict[str, str]:
+        return {"name": self.name, "owner": self.owner, "value": str(self.value)}
```
This should probably be merged into the previous commit as you are using it there...

simon-ess commented on server/recceiver/interfaces.py:
```diff
@@ -21,6 +24,24 @@ class ITransaction(Interface):
     """)
 
 
+@dataclass
+class SourceAddress:
+    host: str
+    port: int
+
+
+@dataclass
+class CommitTransaction:
+    source_address: SourceAddress
+    client_infos: Dict[str, str]
+    records_to_add: Dict[str, Tuple[str, str]]
+    records_to_delete: Set[str]
+    record_infos_to_add: Dict[str, Dict[str, str]]
+    aliases: Dict[str, List[str]]
+    initial: bool
+    connected: bool
+
+
```
This commit seems odd on its own. You create two dataclasses, one of which is not used and one of which is only used as a type hint. Should this be its own commit?

jacomago replied:
This is only for type hinting, in reality the object passing is an ITransaction, but then you don't get the full type info. And I have a feeling it might mess up other things to touch it.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -251,11 +259,11 @@ def _commitWithThread(self, transaction: CommitTransaction):
+            recordInfo[record_id] = RecordInfo(pvName=record_name, recordType=None, infoProperties=[], aliases=[])
             if self.cf_config.recordType:
-                recordInfo[record_id]["recordType"] = record_type
+                recordInfo[record_id].recordType = record_type
```
This feels ugly somehow, but I'm not sure what to do about it.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -267,9 +275,8 @@ def _commitWithThread(self, transaction: CommitTransaction):
             recinfo_wl = [p for p in self.record_property_names_list if p in record_infos_to_add.keys()]
             if recinfo_wl:
-                recordInfo[record_id]["infoProperties"] = list()
                 for infotag in recinfo_wl:
-                    recordInfo[record_id]["infoProperties"].append(
+                    recordInfo[record_id].infoProperties.append(
                         create_property(owner, infotag, record_infos_to_add[infotag])
                     )
 
```
This conditional can be removed now since the "infoProperties" field already exists in the dataclass.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -329,17 +332,17 @@ def _commitWithThread(self, transaction: CommitTransaction):
-                if record_name in recordInfoByName and "aliases" in recordInfoByName[record_name]:
-                    for alias in recordInfoByName[record_name]["aliases"]:
+                if record_name in recordInfoByName:
+                    for alias in recordInfoByName[record_name].aliases:
```
SO MUCH CLEANER

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -525,11 +528,13 @@ def __updateCF__(
+                        if cf_channel["name"] in recordInfoByName:
+                            for alias_name in recordInfoByName[cf_channel["name"]].aliases:
+                                # TODO Remove? This code couldn't have been working....
+                                alias_channel = {"name": alias_name, "properties": [], "owner": ""}
```
Don't we have a dataclass for this yet?

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -525,11 +528,13 @@ def __updateCF__(
+                            for alias_name in recordInfoByName[cf_channel["name"]].aliases:
+                                # TODO Remove? This code couldn't have been working....
+                                alias_channel = {"name": alias_name, "properties": [], "owner": ""}
+                                if alias_name in channels_dict:
```
Is this the part that "couldn't have been workign"? That is, the switch from whatever the `alias` thingy was before (evidently a dictionary) to a string?

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -671,10 +678,10 @@ def __updateCF__(
-            newProps.append(create_recordType_property(owner, recordInfoByName[channel_name]["recordType"]))
-        if channel_name in recordInfoByName and "infoProperties" in recordInfoByName[channel_name]:
-            newProps = newProps + recordInfoByName[channel_name]["infoProperties"]
+        if cf_config.recordType and channel_name in recordInfoByName and recordInfoByName[channel_name].recordType:
```
Is this an edge case that you have found?

jacomago replied:
Was trying to get the type checker happy with the next line.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -79,6 +79,22 @@ class RecordInfo:
+    recordDesc = enum.auto()
+    caPort = enum.auto()
+    pvaPort = enum.auto()
+    time = enum.auto()
```
I can see the advantage here of getting rid of magic strings, but aren't properties somewhat mutable? That is, people can add their own properties above and beyond this default list.

jacomago replied:
Yes, this is just for default properties. Not for instance the environment properties or the info tag properties.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -338,25 +355,18 @@ def _commitWithThread(self, transaction: CommitTransaction):
 
         if transaction.initial:
             """Add IOC to source list """
```
I am finding these """ comments very distracting. Who wrote these????

anderslindho replied:
FWIW they are not comments, just random unbound string literals 🤷

simon-ess replied:
Yes, but they are  clearly used as comments which is super annoying.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -478,41 +477,26 @@ def create_time_property(owner: str, time: str) -> CFProperty:
     # Consider making this function a class methed then 'processor' simply becomes 'self'
     client = processor.client
-    channels_dict = processor.channel_dict
+    channels_iocs = processor.channel_dict
```
I don't get this name change.

jacomago replied:
So channels_dict is actually a dictionary mapping a channel to a list of iocs that the channel is associated to. Tried to make clearer.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -531,18 +515,18 @@ def __updateCF__(
-                if cf_channel["name"] in channels_dict:
-                    cf_channel["owner"] = iocs[channels_dict[cf_channel["name"]][-1]]["owner"]
+                if cf_channel["name"] in channels_iocs:
+                    cf_channel["owner"] = iocs[channels_iocs[cf_channel["name"]][-1]].owner
```
This is an improvement, but UGH

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -531,18 +515,18 @@ def __updateCF__(
                             cf_channel["properties"].append(
                                 create_recordType_property(
-                                    owner, iocs[channels_dict[cf_channel["name"]][-1]]["recordType"]
+                                    ioc_info.owner, iocs[channels_iocs[cf_channel["name"]][-1]]["recordType"]
```
Should this be a `.recordType` here? We use `.owner` above on the same object...

(These objects are very confusing to follow... this is clearly going to be an improvement but ugh)

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -782,15 +775,19 @@ def create_properties(owner, iocTime, recceiverid, hostName, iocName, iocIP, ioc
+    ioc_info: IocInfo, recceiverid: str, channels_iocs: Dict[str, List[str]], iocs: Dict[str, IocInfo], cf_channel
+):
+    channel_name = cf_channel["name"]
+    last_ioc_info = iocs[channels_iocs[channel_name][-1]]
```
Maybe use a similar definition above?

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -515,83 +514,6 @@ def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo]
                 len(new_channels) == 0 or cf_channel["name"] in records_to_delete
             ):  # case: empty commit/del, remove all reference to ioc
                 _log.debug("Channel {s} exists in Channelfinder not in new_channels".format(s=cf_channel["name"]))
```
I am uncomfortable with the ? next to impossible code in the commit message. Can you clarify why this is impossible to hit code?

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -493,38 +489,34 @@ def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo]
 
     if iocid not in iocs:
-        _log.warning("IOC Env Info not found: {iocid}".format(iocid=iocid))
+        _log.warning("IOC Env Info %s not found in ioc list: %s", ioc_info, iocs)
```
Is it possible that this will result in giant logs? If we are printing every IOC here...

anderslindho replied:
Can we _please_ improve this msg/phrasing while we're at it? I _hate_ "IOC Env Info"

simon-ess reviewed (commented):
Overall this is a clear improvement!

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -39,16 +37,50 @@
+    timezone: str = ""
+    findSizeLimit: int = 10000
+
+    @staticmethod
```
Shouldn't this be a class method?

I'd also suggest changing the name to drop "load"; maybe just `from_config_file` 🔧

anderslindho replied:
This comment seems to still apply

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -525,11 +528,13 @@ def __updateCF__(
-                                    alias["properties"] = __merge_property_lists(
+                        if cf_channel["name"] in recordInfoByName:
+                            for alias_name in recordInfoByName[cf_channel["name"]].aliases:
+                                # TODO Remove? This code couldn't have been working....
```
Is there a task or GH issue to track this?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -604,20 +610,21 @@ def __updateCF__(
-                                        alias,
+                                        alias_channel,
                                         processor.managed_properties,
                                     )
```
Why not do this in the above row when you're declaring the dict? 🔧

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -79,6 +79,22 @@ class RecordInfo:
 
 
+class CFPropertyName(enum.Enum):
+    hostName = enum.auto()
```
IMO just import `Enum` and `auto` and skip the namespaces here 🔧

jacomago replied:
Eh, I prefer the namespace for enum.

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -100,13 +100,28 @@ class PVStatus(enum.Enum):
 
 
+@dataclass
+class IocInfo:
```
Preferential, but maybe do `IOCInfo` 🔧

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -100,13 +100,28 @@ class PVStatus(enum.Enum):
+
+    @property
+    def ioc_id(self):
+        return self.hostname + ":" + self.port
```
Maybe drop the setter?

jacomago replied:
setter?

anderslindho replied:
https://docs.python.org/3/library/functions.html#property.setter

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -736,22 +727,24 @@ def __updateCF__(
-    _log.info("Total channels to update: {nChannels} {iocName}".format(nChannels=len(channels), iocName=iocName))
+    _log.info(
+        "Total channels to update: {nChannels} {iocName}".format(nChannels=len(channels), iocName=ioc_info.iocname)
+    )
```
Convert to C-style strings for performance since you're touching this anyway? (for the ones you do touch) 🔧

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -493,38 +489,34 @@ def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo]
 
     if ioc_info.hostname is None or ioc_info.iocname is None:
-        raise Exception("missing hostName or iocName")
+        raise Exception(f"Missing hostName {ioc_info.hostname} or iocName {ioc_info.iocname}")
```
It's f-strings so you can just hello `{ioc_info.hostname=}` 🔧 

**Edit:** Urgh, we're on an ancient python version, so maybe not

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -585,9 +506,96 @@ def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo]
+):
+    for cf_channel in old_channels:
+        if (
+            len(new_channels) == 0 or cf_channel.name in records_to_delete
```
No reason not to do `not new_channels or` ...

anderslindho replied:
Still seems to apply

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -645,6 +653,57 @@ def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo]
+    return alias
+
+
+def __updateCF__(processor: CFProcessor, recordInfoByName: Dict[str, RecordInfo], records_to_delete, ioc_info: IocInfo):
```
We should avoid imitating dunder methods - just do `__updateCF()`

jacomago committed:
Use set for cf_properties

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,25 +136,22 @@ def _startServiceWithLock(self):
-                    record_property_names_list = []
-                if self.conf.getboolean("recordDesc"):
+
+                record_property_names_list = [s.strip(", ") for s in self.cf_config.info_tags.split()]
```
Comment above this row showing example entries would be nice 🔧

jacomago replied:
Added to demo.conf and full.conf

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -71,6 +71,16 @@ def from_config_adapter(conf: ConfigAdapter) -> "CFConfig":
+    value: Optional[str] = None
+
+    def as_dict(self) -> Dict[str, str]:
+        return {"name": self.name, "owner": self.owner, "value": str(self.value)}
```
Given `value: Optional` I'd say this should be
```python
    def as_dict(self) -> Dict[str, Optional[str]]:
        return {"name": self.name, "owner": self.owner, "value": str(self.value)}
```
and you're probably better off doing `"value": self.value or ""` (to avoid the string `"None"`)

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -430,7 +420,7 @@ def create_channel(name: str, owner: str, properties: List[CFProperty]):
 
 
-def create_property(owner: str, name: str, value: str) -> CFProperty:
+def create_property(owner: str, name: str, value: Optional[str]) -> CFProperty:
```
Does this not have any implications on the `CFProperty` ctor? There is no associated change to that class in this commit.

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -474,93 +472,114 @@ def create_time_property(owner: str, time: str) -> CFProperty:
-    client = processor.client
-    channel_ioc_ids = processor.channel_ioc_ids
-    iocs = processor.iocs
-    cf_config = processor.cf_config
-    recceiverid = processor.cf_config.recceiver_id
-    new_channels = set(recordInfoByName.keys())
-    iocid = ioc_info.ioc_id
-
-    if iocid not in iocs:
-        _log.warning("IOC Env Info %s not found in ioc list: %s", ioc_info, iocs)
+def fetch_existing_channels(
+    new_channels: Set[str], client: ChannelFinderClient, cf_config: CFConfig, processor: CFProcessor
+) -> Dict[str, CFChannel]:
+    """A dictionary representing the current channelfinder information associated with the pvNames"""
+    existingChannels = {}
 
-    if ioc_info.hostname is None or ioc_info.iocname is None:
-        raise Exception(f"Missing hostName {ioc_info.hostname} or iocName {ioc_info.iocname}")
+    """
+    The list of pv's is searched keeping in mind the limitations on the URL length
+    The search is split into groups to ensure that the size does not exceed 600 characters
+    """
```
Let's not add further throwaway string literals here - use a comment instead please

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -474,93 +472,114 @@ def create_time_property(owner: str, time: str) -> CFProperty:
+    for channel_name in new_channels:
+        if not searchString:
+            searchString = channel_name
+        elif len(searchString) + len(channel_name) < 600:
```
magic variable

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -474,93 +472,114 @@ def create_time_property(owner: str, time: str) -> CFProperty:
+    The search is split into groups to ensure that the size does not exceed 600 characters
+    """
+    searchStrings = []
+    searchString = ""
```
What's up with the Java style naming all of a sudden?

anderslindho reviewed (commented):
I'll end the review here, as it looks unfinished from 2add2302bc99fa8bbbac98c8404272d1137e2505

jacomago committed:
rename to loads and classmethod
Config.loads

jacomago committed:
Update pydoc, types and comments

Set """ comments as comments
Add pydoc to every method
Add missing types for methods

jacomago committed:
Use static CFProperty methods

In favour of create_property_blah

jacomago committed:
Add constant for search strings

jacomago committed:
Pull out channel_is_old

jacomago committed:
Pull out orphan channel

jacomago committed:
pull out handle channel old and new

jacomago committed:
pull out get_existing channels

jacomago committed:
pull out handle old channels

jacomago committed:
pull out update_existing_channel_diff_iocid

jacomago committed:
pull out create_new_channel

jacomago committed:
Remove extraneous recordType check

jacomago committed:
snake case record type property

jacomago committed:
Use pvstatus enum

snake case pv_status

jacomago committed:
snake case recordinfo

jacomago committed:
static method from channelfinder_dict

jacomago committed:
snake case current time

jacomago committed:
snake case start service with lock

jacomago committed:
snake case stop_service_with_lock

jacomago committed:
snake case _commit_with_lock

jacomago committed:
snake case to_record_infos

jacomago committed:
snake case record_infos

jacomago committed:
record_info like variables snake case

jacomago committed:
existing channels snake case

jacomago committed:
alias_properties rename

jacomago committed:
new_properties snake case

jacomago committed:
updateCF not dunder

jacomago committed:
Make sure every raise has debug info

jacomago committed:
not dunder update cf

jacomago committed:
empty string rather than "None"

jacomago committed:
Update infotags examples

jacomago committed:
remove error format strings

jacomago commented:
For ESS folks, solves CHFI-192 and 198

tynanford commented on server/demo.conf:
```diff
@@ -66,7 +66,7 @@
 # cf-store application
 # a space-separated list of infotags to set as CF Properties
-#infotags = archive foo bar blah
+#infotags = archive, foo, bar, blah
```
are the commas required?

jacomago replied:
Ah, no. Good point. In fact "x,y z"  would give ["x,y", "z"]

tynanford replied:
if they aren't required then I would remove them to avoid confusion or change the comment above to say space or comma separated

tynanford commented:
as the resident ancient python supporter, dataclasses are >= 3.7 and are backported to 3.6 via the pip package. That is fine with me but I thought it is worth mentioning for the 3.6 crowd

anderslindho commented:
> as the resident ancient python supporter, dataclasses are >= 3.7 and are backported to 3.6 via the pip package. That is fine with me but I thought it is worth mentioning for the 3.6 crowd

I think this is an interesting topic, @tynanford: should we continue to support EOL python versions for our new releases?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+    """PV Status values."""
+
+    Active = enum.auto()
+    Inactive = enum.auto()
```
Enum states are typically ALL_CAPS

jacomago replied:
That's true, but I'm mostly using it to get the names as strings. So I would have to have

```
class PVStatus(enum.StrEnum):
   ACTIVE = "Active"
   INACTIVE = "Inactive"
```

anderslindho replied:
Can also do `.title()`:
```python
>>> import enum
>>> class PVStatus(enum.StrEnum):
...    ACTIVE = enum.auto()
...    INACTIVE = enum.auto()
...
>>> PVStatus.ACTIVE.name.title()
'Active'
```

anderslindho replied:
Thinking more about this, `StrEnum` is def the way to go - don't think we have any other option for `CFPropertyName`.

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+    username: str = "cfstore"
+    recceiver_id: str = RECCEIVERID_DEFAULT
+    timezone: Optional[str] = None
+    cf_query_limit: int = 10000
```
Prefer this as global var

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+            username=conf.get("username", "cfstore"),
+            recceiver_id=conf.get("recceiverId", RECCEIVERID_DEFAULT),
+            timezone=conf.get("timezone", ""),
+            cf_query_limit=conf.get("findSizeLimit", 10000),
```
Prefer fallback as global var

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+        """Convert to dictionary for Channelfinder API."""
+        return {"name": self.name, "owner": self.owner, "value": self.value or ""}
+
+    @staticmethod
```
As mentioned before, these type of alternate constructors are typically **class**methods (and then you can also replace `return CFProperty(...)` with `return cls(...)`, but that's optional). A static method is usually just a namespaced method with a loose relationship to the cls.

Same comment applies to all the ones following this

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+    """Information about an IOC instance."""
+
+    host: str
+    hostname: str
+    ioc_name: str
+    ioc_IP: str
```
This seems a bit odd. AFAIK recsync with its native definitions does `hostName`, `iocId`. How come you went for `ioc_IP`?

I don't think IOC-name is a concept outside of ESS?

jacomago replied:
*shrug* 
ioc_name has been in the module a long time.

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+            "properties": [p.as_dict() for p in self.properties],
+        }
+
+    @staticmethod
```
class method

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,244 @@
+        self.current_time: Callable[[Optional[str]], str] = get_current_time
+        self.lock: DeferredLock = DeferredLock()
 
     def startService(self):
```
Should we come back to rename remaining camelCase'd methods/functions later on? 💭

jacomago replied:
This one overrides Twisted.Service.startService so can't be renamed. I'll add a comment.

anderslindho replied:
https://docs.python.org/3/library/typing.html#typing.override - better than a comment IMO

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,244 +310,281 @@ def _startServiceWithLock(self):
-                )
-                continue
-            recordInfoByName[info["pvName"]] = info
+    def record_info_by_name(self, record_infos: Dict[str, RecordInfo], ioc_info: IocInfo) -> Dict[str, RecordInfo]:
```
I think this maybe should be a static method ?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,244 +310,281 @@ def _startServiceWithLock(self):
+                        for record_aliases in record_info_by_name[record_name].aliases:
+                            self.remove_channel(record_aliases, iocid)
+
+    def _commitWithThread(self, transaction: CommitTransaction):
```
Didn't wanna rename also this?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,244 +310,281 @@ def _startServiceWithLock(self):
+            ),
+            time=self.current_time(self.cf_config.timezone),
+            port=transaction.source_address.port,
+            channelcount=0,
```
Could/should this be made default?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,244 +310,281 @@ def _startServiceWithLock(self):
-        if self.iocs[iocid]["channelcount"] == 0:
+            self.iocs[iocid].channelcount -= 1
+        if self.iocs[iocid].channelcount == 0:
             self.iocs.pop(iocid, None)
```
Why default `None` after we've already asserted it exists (we would get an exception on the access otherwise)?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +599,688 @@ def clean_service(self):
+                    [
+                        (CFPropertyName.pvStatus.name, PVStatus.Active.name),
+                        (CFPropertyName.recceiverID.name, recceiverid),
+                    ],
```
kwarg would be nice 🔧

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +599,688 @@ def clean_service(self):
+                    _log.debug("Add existing alias %s to previous IOC: %s", alias_channel, last_alias_ioc_id)
+
+
+def orphan_channel(
```
The naming of all of these functions feels a bit odd.

Also, should they all just be module level methods?

jacomago replied:
Probably not. But a lot in this refactor.

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +599,688 @@ def clean_service(self):
+    return existing_channels
+
+
+def handle_old_channels(
```
shouldn't this be something like just `handle_channels`, seeing as there's one branch that handles both old and new?

anderslindho reviewed (commented):
Some of the old comments haven't yet been addressed, so please go over those and cross-check/follow-up.

I'd like some of the "function or method or classmethod or staticmethod" stuff to be addressed. I think we need to follow this up with an architectural redesign of modules(+subpackages), because this is already confused and getting worse.

Not sure if we should try to fix all of sonar's complaints now or accept as technical debt to be addressed later.

jacomago committed:
default query limit

jacomago committed:
CFProperty use classmethods

jacomago committed:
CFChannel classmethod

jacomago committed:
rename handle_old_channels to handle_channels

jacomago committed:
simpler if case

jacomago committed:
keyword args in using prepareFindArgs

jacomago committed:
Remove unnecessary None return

jacomago committed:
default for channelcount

jacomago committed:
snake case commit with thread

jacomago committed:
Add comments for overriden methods

jacomago committed:
_commit_with_lock submethods to snake case

jacomago committed:
static record_info_by_name

jacomago committed:
reduce passing processor in static methods

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,249 @@
 from twisted.internet.threads import deferToThread
 
 from . import interfaces
+from .interfaces import CommitTransaction
+from .processors import ConfigAdapter
```
We should prob convert the project to use abs imports - could be done together with splitting up `cfstore.py` 💭

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,249 @@
+    cf_query_limit: int = DEFAULT_QUERY_LIMIT
+
+    @classmethod
+    def loads(cls, conf: ConfigAdapter) -> "CFConfig":
```
`loads` usually means "load from **s**tring" (and then `load` is typically from file), but I guess this is OK because `from_configadapter` would be a bit odd/is not immediately clear (that class is a bit strange TBH)

simon-ess replied:
I do agree that I expect `loads` to be from a string, but I would otherwise interpret `load` as just kind of _generic_. So maybe it would be better in this case?

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,249 @@
+        return {"name": self.name, "owner": self.owner, "value": self.value or ""}
+
+    @classmethod
+    def from_channelfinder_dict(cls, prop_dict: Dict[str, str]) -> "CFProperty":
```
No need to specify `channelfinder` when the dict keys maps exactly to members here IMO; now you're also breaking the `as_dict` -> `from_dict` symmetry

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,249 @@
+    aliases: List[str] = field(default_factory=list)
+
+
+class CFPropertyName(enum.Enum):
```
`StrEnum` 🤷

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,249 @@
+        }
+
+    @classmethod
+    def from_channelfinder_dict(cls, channel_dict: Dict[str, Any]) -> "CFChannel":
```
Same as before: should just be `from_dict`

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -103,244 +315,285 @@ def _startServiceWithLock(self):
                 # Are any whitelisted properties not already present on CF?
                 # If so, add them too.
-                properties.update(set(record_property_names_list) - set(cf_properties))
+                properties.update(set(record_property_names_list) - cf_properties)
```
Could just make `record_property_names_list` a set at creation

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +608,687 @@ def clean_service(self):
-    if processor.cancelled:
-        raise defer.CancelledError()
+    if ioc_info.hostname is None or ioc_info.ioc_name is None:
+        raise Exception(f"Missing hostName {ioc_info.hostname} or iocName {ioc_info.ioc_name}")
```
Prefer explicit exception. Can maybe use standard lib `AttributeError` or `ValueError` here, if you don't want to create custom one(s).

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +608,687 @@ def clean_service(self):
-) -> List[Dict[str, str]]:
-    """
-    Merges two lists of properties ensuring that there are no 2 properties with
+    newProperties: List[CFProperty], channel: CFChannel, managed_properties: Set[str] = set()
```
Might as well make this PEP compliant while you're it (-> `new_properties`) - I doubt anything calls it by kwarg

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +608,687 @@ def clean_service(self):
-    for oldProperty in channel["properties"]:
-        if oldProperty["name"] not in newPropNames and (oldProperty["name"] not in managed_properties):
+    newPropNames = [p.name for p in newProperties]
+    for oldProperty in channel.properties:
```
Also here

anderslindho commented on server/recceiver/cfstore.py:
```diff
@@ -355,460 +608,687 @@ def clean_service(self):
 
-def prepareFindArgs(conf, args, size=0):
-    size_limit = int(conf.get("findSizeLimit", size))
+def prepareFindArgs(cf_config: CFConfig, args, size=0) -> List[Tuple[str, str]]:
```
Also this function name

anderslindho reviewed (commented):
It's pretty good, but IMO we should be fixing these remaining idiomatic python/PEP-8 issues before merge. Can all be in a single new commit if you want, TBH.

jacomago committed:
from dict not from_cf_dict

jacomago committed:
more snake case

jacomago committed:
pvstatus to strenum

jacomago committed:
cfpropertyname as strenum

jacomago committed:
record property names list set

jacomago committed:
Specific error class

jacomago commented:
@anderslindho I dropped the override commit, only supported on 3.12 and above.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,248 @@
+        return {"name": self.name, "owner": self.owner, "value": self.value or ""}
+
+    @classmethod
+    def from_dict(cls, prop_dict: Dict[str, str]) -> "CFProperty":
+        """Create CFProperty from Channelfinder json output.
+
+        Args:
+            prop_dict: Dictionary representing a property from Channelfinder.
+        """
+        return cls(
+            name=prop_dict.get("name", ""),
+            owner=prop_dict.get("owner", ""),
+            value=prop_dict.get("value"),
+        )
```
Do you really need this function? You can also allow defaults in the class definition and just do
```py
CFProperty(**some_dict)
```

jacomago replied:
Its the frustrating 'channels' field in the prop_dict :( that blocks doing it that way.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -20,37 +19,248 @@
+            "properties": [p.as_dict() for p in self.properties],
+        }
+
+    @classmethod
+    def from_dict(cls, channel_dict: Dict[str, Any]) -> "CFChannel":
+        """Create CFChannel from Channelfinder json output.
+
+        Args:
+            channel_dict: Dictionary representing a channel from Channelfinder.
+        """
+        return cls(
+            name=channel_dict.get("name", ""),
+            owner=channel_dict.get("owner", ""),
+            properties=[CFProperty.from_dict(p) for p in channel_dict.get("properties", [])],
+        )
```
Same question as above.

simon-ess commented on server/recceiver/cfstore.py:
```diff
@@ -61,39 +271,40 @@ def startService(self):
-                    "iocid",
-                    "iocIP",
-                    RECCEIVERID_KEY,
+                    CFPropertyName.HOSTNAME.value,
+                    CFPropertyName.IOC_NAME.value,
+                    CFPropertyName.IOC_ID.value,
+                    CFPropertyName.IOC_IP.value,
+                    CFPropertyName.PV_STATUS.value,
+                    CFPropertyName.TIME.value,
+                    CFPropertyName.RECCEIVER_ID.value,
```
Do we actually need the `.value`s here? Can we not just work with the enum entities?

simon-ess replied:
This is kind of a general question tbh - everywhere I saw the enums used they were used as their `.value` version. Can't they be let be _except_ when you are printing them?

jacomago replied:
CFPropertyName is a tricky case as you mentioned, in that this is the 'default' property names. You can have them for the env variables and the info tags.

tynanford commented:
> > as the resident ancient python supporter, dataclasses are >= 3.7 and are backported to 3.6 via the pip package. That is fine with me but I thought it is worth mentioning for the 3.6 crowd
> 
> I think this is an interesting topic, @tynanford: should we continue to support EOL python versions for our new releases?

Personally I think there is value supporting 3.6 for the user community. It is EOL but it is the default python version on redhat and the RHEL variants. Redhat backports security fixes to 3.6 and will until 2029. Not saying we need to continue 3.6 support until then but I also don't see a big reason to drop it now

jacomago commented:
@anderslindho name to value in https://github.com/ChannelFinder/recsync/compare/26d6654a5786ff342ad5928093d30ab2e37cd0ab..f74749eb22ba3c1f833fb382dbd903f3f8c7b865 is the only change

tynanford commented:
It looks like StrEnum was introduced in python 3.11 so this breaks for all python versions before then

jacomago commented:
> It looks like StrEnum was introduced in python 3.11 so this breaks for all python versions before then

Looks like we could use https://pypi.org/project/StrEnum/ but it still wouldn't support 3.6...
